### ISSUE TITLE
Research Canvas data - Points data instead of percentages 

### ISSUE THREAD

**Issue by jennlove-um** [tl-its-umich-edu/my-learning-analytics#769]
## Describe your problem or feature you'd like added
For more gamified courses, it would be better to show the Grade Distribution and Assignment Planning views.  Instructors for these types of courses emphasize points rather than grading by percentages earned.

## Describe the solution you'd like
Get grade and assignment data by points instead of by percentages from Canvas data.  Design the Grade Distribution and Assignment Planning alternative views to use points instead of percentages.  This would need to be a parameter that could be configured on the course level.

The first step is to research whether Canvas provides the data that we would need.

**Comment by jonespm** [comment]
It actually *looks* like most courses are graded on points. I wonder why we went with percentages. I think this option sounds like a good one. Canvas provides the data of what the type is displayed but I feel like internally they still use points. . .

|grading_type|% of courses|
|---|---|
|letter_grade|2.06|
|percent|2.56|
|points|81.91|
|gpa_scale|0.03|
|not_graded|5.12|
|pass_fail|8.33|

**Comment by pushyamig** [comment]
Yes, I think we should ask Canvas to make the point grade available in UDW. since it is right there sitting in one of their tables.

**Comment by jonespm** [comment]
I'm not sure what you mean, submission_dim has score in points. We're calculating the percent toward final grade locally. 

What I think this is describing is instead of having percent toward final grade we'd have points attained toward maximum points.

**Comment by jonespm** [comment]
For instance in a gamified course, you want to show something like

600/5000 (points)
This is basically the same as showing
12% of 100% (which is what we're doing now)

But it's showing the points instead of percent on the bar.

**Comment by simsmat116** [comment]
Hello, I am currently in EECS481 and would like to contribute to working on this enhancement for my contribution project. Would anyone be able to give some guidance for this task?

**Comment by justin0022** [comment]
Hi @simsmat116, I think this is a fairly large feature with many different underlying components. Is there a specific component you'd be interested in implementing? I imagine there is considerable work to be done both in the front and back-end - do you have a preference?

**Comment by simsmat116** [comment]
Hi @justin0022 ,  would I be able to work on the back-end relating the Grade Distribution part of this issue?  I could also work on the front-end component for this as well depending on how long it takes to work on the back-end. I am looking to spend at least 12 hours of designing the changes, implementing them and testing them. Would you say this would be reasonable? Thanks!

**Comment by justin0022** [comment]
I think that's reasonable. What do you think @jonespm & @pushyamig?

**Comment by jonespm** [comment]
I was talking to @jennlove-um  about this the other day and I'm not 100% sure of the desire here. This is the Canvas data dictionary: https://portal.inshosteddata.com/docs

The assignment_dim defines grading_type (which we aren't currently pulling in) however everything is still always in points_possible. I think MyLA is changing this over to a percentage scale. I didn't write the code for this view so not 100% sure how it was working and would have to research it myself. ;)
I "think" all the data is there already unless we also need to populate grading_type, though I feel like MyLA's grading_type may differ from the type in Canvas. So we probably need a few things.

Backend:
- A way to set the course specific grading type in MyLA. I'd probably add it as a new column on the course table and have it configurable from the admin. The course table is a little unusual as some values are populated from the cron and others are populated from the admin. Really we should split up these tables but let's say we have a new column myla_grading_type that would be defined in the UI and also be varchar (percent, points, both). I think we could probably do both.

- Eventually we'll want instructor to be able to configure this but that's a future feature, we might also need to pull grading_type from cron

Frontend:
So now once that's configured if we have percent configured here the UI would be the same as it is now. However if we have points it would display the top bar as points rather than percentages. I also think the instructor may want to have both I believe most of the code for that is going to be in the React 
component assets/src/components/d3/createProgressBar.js

I'm really not sure how points data would push down into the lower "Percent of final grade". I guess that would also have to change but maybe not?

These are just for the assignment planning. Grade distribution might be a little easier and just show the raw points there instead of the calculated percentage. Maybe that's the value @pushyamig was referring to earlier that doesn't seem to be available from Canvas?

**Comment by simsmat116** [comment]
@jonespm I was working on creating the new column in the course table and was not sure how to ensure that this additional column will be captured. Is there a database schema file that I should edit or should I go about this a different way? Thanks!

Also, I have the field configurable from the admin portal and have also made sure that the back-end response for assignment planning is capturing the grading_type for the specified course. Now I believe that I will just need to make changes to the front-end as the point data and percentage data are already captured in the back-end response.

**Comment by jonespm** [comment]
Yeah, it can be hard to test issues relating to database changes because we pull our data from a source that doesn't have a test instance.

There's no schema file you need to edit, just change the model.py, execute `manage.py makemigrations` in the container and that should write out the migration.

You'd just have to write the cron code that would populate it and we'll re-generate the sample SQL. For testing you can just write SQL that will populate this column and add it to your own copy of the sample. 

We're hoping to make this process better by having actual fixtures in the source but that won't happen until next semester.

**Comment by simsmat116** [comment]
@jonespm Ok I ran the migration and it seems to have worked. I was populating the field for testing via the admin portal. Is this alright or should I go ahead and write some SQL to populate this? 

Also, I currently have the myla_grading_type field options as (Percentage, Points, or Both) and was wondering if Both is necessary for the changes I am making. I was able to display two ProgressBars on the AssignmentPlanning view, when the grading type == BOTH, but think that it may be a bit more difficult for the GradingDistribution as two histograms would take up a large amount of space on the page. I've included a screenshot below of what the Assignment Planning looks like with both progress bars:
![Screen Shot 2019-11-26 at 2 33 10 PM](https://user-images.githubusercontent.com/46581282/69739323-827d9b80-1105-11ea-9387-9f9567fdfcf7.png)

 What are you thoughts on having only (Percentage and Points) as the grading type options for now?

**Comment by zqian** [comment]
I think the display is either by Percentage or by Points, not both. Admin can set the option for course, since we do not support instructor settings, yet.

**Comment by simsmat116** [comment]
@zqian I have set up the Percentage display and Point display for both the Grade Distribution and Assignment Planning views. I have also written tests for these as well. I was wondering what the expectation was relating to actually populating the new field (myla_grading_type) in the _course_ table in the database. Should I write something in the cron job (such as just populating all course to display data as percentages) to do this or should I not worry about this?

**Comment by pushyamig** [comment]
@simsmat116 Regarding the Point vs percentage for Assignment view I would suggest to focus on Grades distributions due to the intricacy  of that assignment view and current [redesign](https://github.com/tl-its-umich-edu/my-learning-analytics/pull/702) going of it. This is a research issues so if your focus is grades view then will can be simple

**Comment by simsmat116** [comment]
@pushyamig I have already finished up working on the Assignment Planning view changes, both back-end and front-end, but I can revert these changes if needed. I have also completed the changes needed for the Grade Distribution and was planning on making a pull request in the next day or two. Let me know if I should remove the Assignment Planning changes when making a PR.

**Comment by pushyamig** [comment]
@simsmat116 in that case have 2 separate Pull requests one for grades and another for Assignment   so that way you can still earn points for the effort  you did and we also get see your approch as well as I said this is a reaseach issues any effort don't go waste

**Comment by pushyamig** [comment]
Regarding the `myla_grading_type` new column addition to the `course` table, this is not a solution but the idea. 
1. create a `CharField` column in the course table in `models.py` that takes either `points` or `percent` option in that column, default it to percent
2. From the Admin view for the course expose it such that Admins can configure it to points or percentage. Based on the selection the tables get updated with the value. HTML element can be textbox, dropdown, or 2 checkboxes. just see what is feasible.
3. in the `views.py` grade distribution/assignment functional call pull the points vs percent info from `course` table and generate the JSON data for Front-end. Based on the Json data Frontend knows which format to present the data
